File: src/main/java/cube8540/oauth/authentication/credentials/oauth/client/infra/ClientAPIExceptionTranslator.java
package cube8540.oauth.authentication.credentials.oauth.client.infra;

import cube8540.oauth.authentication.credentials.oauth.client.domain.exception.ClientAuthorizationException;
import cube8540.oauth.authentication.credentials.oauth.client.domain.exception.ClientInvalidException;
import cube8540.oauth.authentication.credentials.oauth.client.domain.exception.ClientNotFoundException;
import cube8540.oauth.authentication.credentials.oauth.client.domain.exception.ClientRegisterException;
import cube8540.oauth.authentication.error.ExceptionTranslator;
import cube8540.oauth.authentication.error.message.ErrorMessage;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;

@Slf4j
@Component
public class ClientAPIExceptionTranslator implements ExceptionTranslator<ErrorMessage<Object>> {
    @Override
    public ResponseEntity<ErrorMessage<Object>> translate(Exception exception) {
        if (exception instanceof ClientNotFoundException) {
            ClientNotFoundException e = ((ClientNotFoundException) exception);
            return response(HttpStatus.NOT_FOUND, ErrorMessage.instance(e.getCode(), e.getMessage()));
        } else if (exception instanceof ClientInvalidException) {
            ClientInvalidException  e = ((ClientInvalidException) exception);
            return response(HttpStatus.BAD_REQUEST, ErrorMessage.instance(e.getCode(), e.getErrors().toArray()));
        } else if (exception instanceof ClientRegisterException) {
            ClientRegisterException e = ((ClientRegisterException) exception);
            return response(HttpStatus.BAD_REQUEST, ErrorMessage.instance(e.getCode(), e.getMessage()));
        } else if (exception instanceof ClientAuthorizationException) {
            ClientAuthorizationException e = ((ClientAuthorizationException) exception);
            return response(HttpStatus.UNAUTHORIZED, ErrorMessage.instance(e.getCode(), e.getMessage()));
        } else {
            log.error("Handle exception {}, {}", exception.getClass(), exception.getMessage());
            return response(HttpStatus.INTERNAL_SERVER_ERROR, ErrorMessage.UNKNOWN_SERVER_ERROR);
        }
    }
}

File: src/main/kotlin/cube8540/oauth/authentication/oauth/client/infra/ClientAPIExceptionTranslator.kt
package cube8540.oauth.authentication.oauth.client.infra

import cube8540.oauth.authentication.oauth.client.domain.ClientAuthorizationException
import cube8540.oauth.authentication.oauth.client.domain.ClientInvalidException
import cube8540.oauth.authentication.oauth.client.domain.ClientNotFoundException
import cube8540.oauth.authentication.oauth.client.domain.ClientRegisterException
import cube8540.oauth.authentication.error.ExceptionTranslator
import cube8540.oauth.authentication.error.message.ErrorMessage
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Component

@Component
class ClientAPIExceptionTranslator: ExceptionTranslator<ErrorMessage<Any>> {

    private val logger = LoggerFactory.getLogger(this.javaClass)

    override fun translate(exception: Exception): ResponseEntity<ErrorMessage<Any>> = when (exception) {
        is ClientNotFoundException -> {
            response(HttpStatus.NOT_FOUND, ErrorMessage.instance(exception.code, exception.message))
        }
        is ClientInvalidException -> {
            response(HttpStatus.BAD_REQUEST, ErrorMessage.instance(exception.code, exception.errors.toTypedArray()))
        }
        is ClientRegisterException -> {
            response(HttpStatus.BAD_REQUEST, ErrorMessage.instance(exception.code, exception.message))
        }
        is ClientAuthorizationException -> {
            response(HttpStatus.FORBIDDEN, ErrorMessage.instance(exception.code, exception.message))
        }
        else -> {
            logger.error("Handle exception {} {}", exception.javaClass, exception.message)
            response(HttpStatus.INTERNAL_SERVER_ERROR, ErrorMessage.UNKNOWN_SERVER_ERROR)
        }
    }
}